### ISSUE TITLE
MNT: migrating to ruff 0.8.0

### ISSUE THREAD

**Issue by neutrinoceros** [astropy/astropy#17430]
Another massive ruff release associated with a [migration guide](https://astral.sh/blog/ruff-v0.8.0) just dropped. Formatting is stable as compared with 0.7.x, but there are a couple newly stabilized rules we need to take into account in order to migrate.
Let me list them here, sorted by decreasing number of violations

## autofix available (tot: 147)
- [RUF022](https://docs.astral.sh/ruff/rules/unsorted-dunder-all/) (123)
    - #17437 
- [RUF021](https://docs.astral.sh/ruff/rules/parenthesize-chained-operators/) (20)
	- #17436 
- [RUF023](https://docs.astral.sh/ruff/rules/unsorted-dunder-slots/) (3)
	- #17435 
- [PYI064](https://docs.astral.sh/ruff/rules/redundant-final-literal/) (1)
	- #17434

## no autofix (tot: 16)
- [UP031](https://docs.astral.sh/ruff/rules/printf-string-formatting/) (9)
    - #17438 *or* #17494
    - #17452
    - https://github.com/astral-sh/ruff/issues/14555
- [PLC0206](https://docs.astral.sh/ruff/rules/dict-index-missing-items/) (7)
	- #17439  
	- #17440
	- #17441
	- #17442
	- #17443

I think the easiest approach would be to have one PR per rule for the ones that can be autofixed. In particular RUFF022 applies very widely but it shouldn't need a thorough review from every subpackage owners. If this is somehow controversial another option is to disable this rule globally or at the subpackage level.

For rules requiring manual refactors, I observe that:
- 8/9 violations for UP031 are in `convolution`, so it should be easy to separate PRs by subpackage
- PLC026 is another story and has 7 violations spread over 5 packages. For this one I think we should disable the rule by subpackage for now. *update: as can be seen above I changed my mind about this seeing how easy it was to actually fix these violations*

**Comment by pllim** [comment]
When are they going to stop changing stuff?

**Comment by mhvk** [comment]
I'd argue to disable `RUF021` - I think it is OK to assume people know that `and` has precedence over `or` - no different from `*` having precedence over `+`.

The others all seem reasonable to fix (though I share @pllim's sigh of having yet more style PRs ahead... so much time spent on it... that said, the precommit stuff does really help keep things uniform, so perhaps it is all worth it).

**Comment by neutrinoceros** [comment]
> When are they going to stop changing stuff?

I don't think we can (or should) assume that they will. We're bound to have this sort of issue with upgrading as long as we configure ruff with all rules + opt outs.

For reference, and I am not arguing against it, astropy is the only project I work with that went this way, and I'm seeing much more stability in projects that use an opt-in configuration style (only include the rule sets you want).

I'm happy to spend some paid time helping with these arguably painful upgrades, but I we may want to re-assess, as a project, if this worth the effort and money in the long run.

**Comment by neutrinoceros** [comment]
> I'd argue to disable RUF021 - I think it is OK to assume people know that `and` has precedence over `or`

Btw, I for one can never remember this, so enforcing this rule would actually help me 😅

**Comment by neutrinoceros** [comment]
I opened and linked 5 PRs for all auto fixes + manual fixes for UP031.
I also opened an issue upstream about the one new UP031 violation from `io.fits` which I think should be ignored https://github.com/astral-sh/ruff/issues/14555

**Comment by mhvk** [comment]
Well, my slight annoyance with RUF021 is amply outweighed by how it makes many lines shorter. And that the expressions linked by `and` are often quite long so that it does actually help to have parenthesizes.  So, I went ahead and merged the PR.

**Comment by nstarman** [comment]
@neutrinoceros can we upgrade now?

**Comment by neutrinoceros** [comment]
Still blocked by:
- https://github.com/astropy/astropy/pull/17438
- https://github.com/astropy/astropy/pull/17494
(these are alternative fixes)
